nixos/modules/services/misc/matrix-appservice-discord.nix

```diff
@@ -0,0 +1,118 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  dataDir = "/var/lib/matrix-appservice-discord";
+in {
+  options = {
+    services.matrix-appservice-discord = {
+      enable = mkEnableOption "a bridge between Matrix and Discord";
+
+      port = mkOption {
+        type = types.int;
+        default = 9005; # from https://github.com/Half-Shot/matrix-appservice-discord/blob/master/package.json#L11
+        description = ''
+          Port number on which the bridge should listen for internal communication with the Matrix homeserver.
+        '';
+      };
+
+      config = mkOption rec {
+        type = types.attrs;
+        apply = recursiveUpdate default;
+        default = {
+          database = {
+            filename = "${dataDir}/discord.db";
+          };
+        };
+        example = literalExample ''
+          {
+            bridge = {
+              domain = "public-domain.tld";
+              homeserverUrl = "http://public-domain.tld:8008";
+            };
+          }
+        '';
+        description = ''
+          <filename>config.yaml</filename> configuration as a Nix attribute set.
+
+          Configuration options should match those described in
+          <link xlink:href="https://github.com/Half-Shot/matrix-appservice-discord/blob/master/config/config.sample.yaml">
+          config.sample.yaml</link>.
+
+          <option>config.bridge.domain</option> and <option>config.bridge.homeserverUrl</option>
+          should be set to match the public host name of the Matrix homeserver for webhooks and avatars to work.
+
+          Secret tokens should be specified using <option>environmentFile</option>
+          instead of this world-readable attribute set.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = ''
+          APPSERVICE_DISCORD_AUTH_CLIENT_I_D="12345"
+          APPSERVICE_DISCORD_AUTH_BOT_TOKEN="foobar"
+        '';
+        description = ''
+          File containing environment variables to be passed to the matrix-appservice-discord service,
+          in which secret tokens can be specified securely.
+        '';
+      };
+
+      homeserverService = mkOption {
+        type = types.string;
+        default = "matrix-synapse.service";
+        description = ''
+          Matrix homeserver service to wait for when starting the application service.
+        '';
+      };
+    };
+  };
+
+  config = let
+    cfg = config.services.matrix-appservice-discord;
+    appDir = "${pkgs.nodePackages.matrix-appservice-discord}/lib/node_modules/matrix-appservice-discord";
+    configJSON = pkgs.writeText "matrix-appservice-discord-config.json" (builtins.toJSON cfg.config);
+    configYAML = pkgs.runCommand "matrix-appservice-discord-config.yaml" { preferLocalBuild = true; } ''
+      ${pkgs.remarshal}/bin/json2yaml -i ${configJSON} -o $out
```

Comment:
Is this one of the tools that are unable to properly parse JSON files? Otherwise I'd skip that step :)

Reply:
YAML is de facto the standard configuration format for all Matrix-related stuff.
I'd like to remove this conversion step but there's no `builtins.toYAML` (yet).

Reply:
Actually, JSON is a subset of YAML, so each JSON string can usually be parsed by a YAML parser (have a look at e.g. the GitLab module where this is heavily used).

However I think that there were some YAML parsers that broke with JSON, so I'm not sure if that works here :)

Reply:
Oh I didn't know that. It seems to work indeed!